### PR TITLE
Arbitrary headers, POST support, POST body support, Content-Type support, linting

### PR DESCRIPTION
**Pull request by mzpqnxow**
Hi, I use gobuster quite a bit as it is a simple and very fast tool for enumerating files and directories. There were some limitations that I wanted to deal with though- specifically, gobuster would not find even the most obviously named API endpoints, despite the method names being in the wordlist. This turned out to be because they required either a different HTTP verb (i.e. POST) or they required a special Content-Type (i.e. application/json)

I've added support for these things, including the ability to specify a POST body on the command-line. If I have a spare few moments I think it would be nice to support curl style `-b @filename.json` to put the POST body contents in a file since the command-line can be a little clunky.

Your README comments on the poor style of the code; I didn't find it to be all that bad, but I did spend about an hour linting it extensively with `gometalinter` - I thought you might appreciate that. You can go through the diffs and spot the cleanups. Many of them were just adding `FunctionName ... <comments>` to the headers of all exported functions, though there were quite a few other legitimate cleanups.

Apologies in advance for not splitting out the commits better, I know this is a large PR

Also note you will likely want to change the import ./libgobuster back to github.com/OJ/gobuster/libgobuster if you accept this PR. Because there were heavy changes to libgobuster and not just main I had to work with the local copy of libgobuster. Then again, I'm not sure it matters

Let me know if you want to accept this, if you don't I will keep a fork for myself

Thx
A

### CODE REVIEW

File: libgobuster/dir.go [hunk 1]
```diff
@@ -43,27 +47,40 @@ func (rh *RedirectHandler) RoundTrip(req *http.Request) (resp *http.Response, er
...
-func MakeRequest(s *State, fullUrl, cookie string) (*int, *int64) {
-	req, err := http.NewRequest("GET", fullUrl, nil)
+// MakeRequest ... Make a request to the given URL
+func MakeRequest(s *State, fullURL, cookie string) (*int, *int64) {
+
+	request, err := http.NewRequest(s.Verb, fullURL, strings.NewReader(s.Body))
 
 	if err != nil {
-		return nil, nil
+		os.Exit(1)
```
Comment (0xdevalias) [inline libgobuster/dir.go]: Not a fan of this as a pattern, ideally all errors should be passed up higher, then handled at the top. Exits/panics deep down make life hard.

I believe this has already been solved as part of https://github.com/OJ/gobuster/pull/56 (still WIP)

File: libgobuster/statehelpers.go [hunk 2]
```diff
@@ -19,7 +19,7 @@ package libgobuster
 import (
 	"crypto/tls"
 	"fmt"
-	"github.com/hashicorp/go-multierror"
+	"go-multierror"
```
Comment (0xdevalias) [inline libgobuster/statehelpers.go]: Why this change? It's not part of stdlib?

File: libgobuster/statehelpers.go [hunk 3]
```diff
@@ -39,6 +40,36 @@ func InitState() State {
 	}
 }
 
+// ValidVerb ...
+func ValidVerb(verb string) bool {
```
Comment (0xdevalias) [inline libgobuster/statehelpers.go]: There is a potential we may want to be able to use non-valid verbs, or verbs that are only valid under certain protocols. Not a big issue for now though.

File: libgobuster/statehelpers.go [hunk 4]
```diff
@@ -78,23 +109,23 @@ func ValidateState(
...
+		if err := validateDirModeState(s, extensions, codes, proxy, errorList); err.ErrorOrNil() != nil {
 			errorList = err
 		}
 	}
 
 	return errorList
 }
 
-func ValidateDirModeState(
+func validateDirModeState(
```
Comment (0xdevalias) [inline libgobuster/statehelpers.go]: Why was this switched to a lowercase?

File: main.go [hunk 5]
```diff
@@ -19,11 +19,11 @@ package main
 import (
 	"flag"
 	"fmt"
-
-	"github.com/OJ/gobuster/libgobuster"
+	"./libgobuster"
```
Comment (0xdevalias) [inline main.go]: This should be changed back I believe, to follow 'proper' standards

**Comment by mzpqnxow** [comment]
Thank you much for the feedback, I'm not experienced at all in golang so it is *really* helpful.. I'm wondering if maybe a gobuster2 would be a better idea rather than merging all into gobuster

File: libgobuster/statehelpers.go (hunk 4 above)
Comment (mzpqnxow) [reply libgobuster/statehelpers.go]: lowercase because it is static context (not meant to be used from another file) per go standards (if I understand correctly)

File: main.go (hunk 5 above)
Comment (mzpqnxow) [reply main.go]: yep

File: libgobuster/statehelpers.go (hunk 2 above)
Comment (mzpqnxow) [reply libgobuster/statehelpers.go]: honestly I have no idea how this happened but it breaks the build.. oops :>

File: libgobuster/statehelpers.go (hunk 3 above)
Comment (mzpqnxow) [reply libgobuster/statehelpers.go]: Yes, just need to provide a clean way for the user to specify if there should be a content length/body for an invalid verb I guess

**Comment by 0xdevalias** [comment]
My personal feeling is that forking and making a 'new gobuster' wouldn't serve much purpose, except for possibly getting faster turnaround time on dev (since @OJ doesn't seem to have much time to dedicate to this), but at the risk of fragmenting the community around the tool. /2c

**Comment by mzpqnxow** [comment]
@OJ no problem. I’ve been extremely busy the last month and haven’t been able to work on it (and I didn’t sufficiently QA it)

Ill rebase to 2.0.0 and first may try to add a test suite in the form of a simple ephemeral web server. I haven’t look at 2.0.0 but I’m looking forward to taking a look

At that point I’ll start to splice the PR content and add it (to 2.0.0.0)

Thanks

**Comment by OJ** [comment]
Hey @mzpqnxow we've got a serious rebuild going on the v3.0-working branch. Would you be able to find time to rebase your efforts onto that instead?

Apologies, trying my best to juggling ALL THE BALLS at the moment.

**Comment by OJ** [comment]
Hey @mzpqnxow sorry to have the earth moving underneath you. We've landed 3.0 now and so things hav changed a bit. Would you be open to rebasing on master now please? We've got support for arbitrary headers via the `-H` flag now.

**Comment by FireFart** [comment]
Closing because of no response. If you want to rebase it feel free to reopen the PR
